jme3-examples/src/main/java/jme3test/terrain/PBRTerrainAdvancedTest.java

```diff
@@ -0,0 +1,428 @@
+        inputManager.addMapping("triPlanar", new KeyTrigger(KeyInput.KEY_P));
+        inputManager.addMapping("toggleNight", new KeyTrigger(KeyInput.KEY_N));
+        
+        inputManager.addListener(actionListener, "triPlanar");
+        inputManager.addListener(actionListener, "toggleNight");
+        
+        
+        keybindingsText = new BitmapText(assetManager.loadFont("Interface/Fonts/Default.fnt"));
+        keybindingsText.setText("Press 'N' to toggle day/night fade (takes a moment) \nPress 'P' to toggle tri-planar mode");
+        getGuiNode().attachChild(keybindingsText);
+        keybindingsText.move(new Vector3f(200,120,0));
+        
+    }
+    private ActionListener actionListener = new ActionListener() {
+
+        public void onAction(String name, boolean pressed, float tpf) {
+            if (name.equals("triPlanar") && !pressed) {
+                triPlanar = !triPlanar;
+                if (triPlanar) {
+                    matTerrain.setBoolean("useTriPlanarMapping", true);
+                    // planar textures don't use the mesh's texture coordinates but real world coordinates,
+                    // so we need to convert these texture coordinate scales into real world scales so it looks
+                    // the same when we switch to/from tr-planar mode (1024f is the alphamap size)
+                    matTerrain.setFloat("AlbedoMap_0_scale", 1f / (float) (1024f / dirtScale));
+                    matTerrain.setFloat("AlbedoMap_1_scale", 1f / (float) (1024f / darkRockScale));
+                    matTerrain.setFloat("AlbedoMap_2_scale", 1f / (float) (1024f / snowScale));
+                    matTerrain.setFloat("AlbedoMap_3_scale", 1f / (float) (1024f / tileRoadScale));
+                    matTerrain.setFloat("AlbedoMap_4_scale", 1f / (float) (1024f / grassScale));
+                    matTerrain.setFloat("AlbedoMap_5_scale", 1f / (float) (1024f / marbleScale));
+                    matTerrain.setFloat("AlbedoMap_6_scale", 1f / (float) (1024f / gravelScale));
+                } else {
+                    matTerrain.setBoolean("useTriPlanarMapping", false);
+
+                    matTerrain.setFloat("AlbedoMap_0_scale", dirtScale);
+                    matTerrain.setFloat("AlbedoMap_1_scale", darkRockScale);
+                    matTerrain.setFloat("AlbedoMap_2_scale", snowScale);
+                    matTerrain.setFloat("AlbedoMap_3_scale", tileRoadScale);
+                    matTerrain.setFloat("AlbedoMap_4_scale", grassScale);
+                    matTerrain.setFloat("AlbedoMap_5_scale", marbleScale);
+                    matTerrain.setFloat("AlbedoMap_6_scale", gravelScale);
+
+                }
+            }
+            if (name.equals("toggleNight") && !pressed) {
+                isNight = !isNight;
+                //ambient and direcitonal light are faded smoothly in update loop below !
+                
+            }
+        }
+    };
+    
+        private float dayLightIntensity = 1.0f;
```

Comment:
Please declare fields *before* methods: `dayLightIntensity`, `nightLightIntensity`, and `actionListener`.
Also, please declare all 3 to be `final`.